ffi/src/sspi/sec_winnt_auth_identity.rs

```diff
@@ -346,7 +349,10 @@ pub unsafe fn auth_data_to_identity_buffers_w(
 
         Ok(CredentialsBuffers::AuthIdentity(AuthIdentityBuffers {
             user,
-            domain: raw_str_into_bytes((*auth_data).domain as *const _, (*auth_data).domain_length as usize * 2),
+            domain: credentials_str_into_bytes(
+                (*auth_data).domain as *const _,
+                (*auth_data).domain_length as usize * 2,
```

Comment:
question: do you know if domain_length is the number of code points or the number of characters of the UTF-16 string? Because some characters may be encoded on multiple code points, causing the maths to be wrong here.

Reply:
We do not convert/reencode the string here. we only read the memory into a buffer (`Vec<u8>`). The number of bytes is `domain_length` * 2, where 2 is the size of the `unsigned short`. 
https://learn.microsoft.com/en-us/windows/win32/api/sspi/ns-sspi-sec_winnt_auth_identity_w
We use `String::from_utf16_lossy` in our codebase when we need the `String`.

So, everything should be okay here

Reply:
I see the following at many places in the docs:

> The length, **in characters**, of the user string, not including the terminating null character.

So what you said:

> The number of **bytes** is domain_length * 2

Does not sound correct to me.

For instance, let’s take the character [𐐷](https://en.wikipedia.org/wiki/%F0%90%90%8F), which is mentioned in the [UTF-16 wikipedia page](https://en.wikipedia.org/wiki/UTF-16) as being encoded on _**2** 16-bit code points_. The string `"𐐷"` is _1_ character, but the number of bytes is 2 × 2 = _4_ which is not equal to domain_length (1 character) × 2 = 1 × 2 = 2.

Do you think you could double check that? If necessary, let’s address that in follow up PRs.

> We use `String::from_utf16_lossy` in our codebase when we need the `String`.

Note that in such cases we should probably use `String::from_utf16` instead. Otherwise, we are silently ignoring invalid code points, which should not be accepted in the first place.

Reply:
FYI, there is a function in the crate `widestring` to handle the situation were you have a pointer to the beginning of a UTF-16 string along its length _in characters_: https://docs.rs/widestring/latest/widestring/ucstring/struct.U16CString.html#method.from_ptr

Reply:
Hmm, that sounds reasonable. I got your idea.
Then we need to introduce a new dependency and fix the code in many places. Let's create a new issue and refactor the code in a separate PR. What do you think?

Reply:
This sounds like the right course of action to me!

Reply:
Let me create an issue

Reply:
https://github.com/Devolutions/sspi-rs/issues/306